### PR TITLE
Don't bother creating taggers when there are no diagnostics to display.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
They're expensive and end up consuming lots of resources.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @amcasey @jasonmalinowski @heejaechang

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 1]
```diff
@@ -53,6 +53,7 @@ public void Dispose()
                 foreach (var kvp in _idToProviderAndTagger)
                 {
                     var tagger = kvp.Value.Item2;
+
```
Comment (amcasey) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: Intentional?

**Comment by amcasey** [comment]
Seems sensible.  I'll try to verify manually.

**Comment by CyrusNajmabadi** [comment]
Jenkins retest this please.

**Comment by amcasey** [comment]
Yep, that seems to prevent hundreds of subscriptions to ```OptionService.OptionChanged``` per InteractiveWindow submission.  Thanks!

**Comment by amcasey** [comment]
:+1:

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: Was refactoring this a bit.  Then decided to revert that part.  Guess a newline snuck in.  :)

**Comment by jasonmalinowski** [comment]
I find your lack of tests disturbing.

Otherwise, this is seems reasonable given the current architecture. I'd be curious if it's possible to improve the architecture eventually to have more state get shared between the different diagnostic taggers.

**Comment by amcasey** [comment]
@KevinH-MS asked me the same question: does each submission need to listen to options changes individually?  Could they stop listening for options changes once they become readonly?

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 2]
```diff
@@ -137,6 +138,15 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
                 ValueTuple<TaggerProvider, IAccurateTagger<TTag>> providerAndTagger;
                 if (!_idToProviderAndTagger.TryGetValue(id, out providerAndTagger))
                 {
+                    // We didn't have an existing tagger for this diagnostic id.  If there are no actual 
+                    // diagnostics being reported, then don't bother actually doing anything.  This saves
+                    // us from creating a lot of objects, and subscribing to tons of events that we don't
+                    // actually need (since we don't even have any diagnostics to show!).
+                    if (e.Diagnostics.Length == 0)
```
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: diagnostic that are reported and diagnostic each tagger cares are different. so, probably need to filter these to see whether there is any diagnostic actually  we care.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: How would we do that?

**Comment by CyrusNajmabadi** [comment]
> @KevinH-MS asked me the same question: does each submission need to listen to options changes individually? Could they stop listening for options changes once they become readonly?

I don't see how.  Even if the submission is readonly, they can still be affected by options.  For example, if a user turns on/off highlight references, then that needs to be respected by the tagger regardless of whether or not the buffer is readonly.

**Pull request merged**